torch/_dynamo/compiled_autograd.py

```diff
@@ -305,18 +305,35 @@ def is_sym_node(self, node):
             in [torch.ops.aten.sym_size.int, torch.ops.aten.sym_numel.default]
         )
 
-    def remove_dead_sym_nodes(self):
-        for node in reversed(list(self.fx_tracer.graph.nodes)):
-            if (
-                node.op == "call_function"
-                and node.target == operator.eq
-                and (self.is_sym_node(node.args[0]) or self.is_sym_node(node.args[1]))
-            ):
-                if len(node.users) == 0:
-                    self.fx_tracer.graph.erase_node(node)
-            if self.is_sym_node(node):
-                if len(node.users) == 0:
-                    self.fx_tracer.graph.erase_node(node)
+    def dce(self):
+        # Most of these removed nodes would have been removed during Dynamo and AOTDispatch
+        # Remove some of these nodes earlier to improve compilation speed
+        impure_targets = {
+            call_hook,
+            call_backward,
+            FakeCompiledAutogradEngine._exec_final_callbacks_stub,
+            torch.ops.inductor.accumulate_grad_.default,
+        }
+
+        # Dynamo guards will error instead of creating aliasing guards unless we unpack them in the graph
+        unpack_nodes: Set[torch.fx.Node] = set()
+        i = len(self.graph_placeholders)
+        for node in self.fx_tracer.graph.nodes:
+            if i == 0:
+                break
+
+            unpack_nodes = unpack_nodes | set(node.users.keys())
+            i -= 1
```

Comment:
Add some asserts about the node types here.  Use OrderedSet to be deterministic.